### PR TITLE
stdenv: warn about use of stdenv.glibc

### PR DESCRIPTION
**Pull request by Artturin**
the todo was added in 
[`119920f` (#34645)](https://github.com/NixOS/nixpkgs/pull/34645/commits/119920faa6717f8075e0ea44553203171a71f9ea)

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Ericson2314** [comment]
Oh glad to see this cleaned up finally!! But I am afraid most of these `stdenv.glibc` should become `stdenv.cc.libc` not `glibc`.

**Comment by Artturin** [comment]
i changed some of them to `stdenv.cc.libc` but most of the place that use this are for unfree programs

File: pkgs/development/python-modules/twisted/default.nix [hunk 1]
```diff
@@ -54,7 +54,7 @@ buildPythonPackage rec {
   # twisted.python.runtime.platform.supportsINotify() == False
   postPatch = lib.optionalString stdenv.isLinux ''
     substituteInPlace src/twisted/python/_inotify.py --replace \
-      "ctypes.util.find_library(\"c\")" "'${stdenv.glibc.out}/lib/libc.so.6'"
+      "ctypes.util.find_library(\"c\")" "'${stdenv.cc.libc}/lib/libc.so.6'"
```
Comment (Mindavi) [inline pkgs/development/python-modules/twisted/default.nix:57]: Also drop the `.6` like with monotonic?

File: pkgs/tools/inputmethods/m17n-db/default.nix [hunk 2]
```diff
@@ -14,8 +14,7 @@ stdenv.mkDerivation rec {
 
   strictDeps = true;
 
-  configureFlags = lib.optional (stdenv ? glibc)
-    "--with-charmaps=${stdenv.glibc.out}/share/i18n/charmaps"
+  configureFlags = [ "--with-charmaps=${glibc.out}/share/i18n/charmaps" ]
```
Comment (Mindavi) [inline pkgs/tools/inputmethods/m17n-db/default.nix:17]: Could this break darwin?

Comment (Mindavi) [reply pkgs/tools/inputmethods/m17n-db/default.nix:17]: Ah, it's a linux-only package.

File: pkgs/tools/system/hardinfo/default.nix [hunk 3]
```diff
@@ -30,7 +30,7 @@ stdenv.mkDerivation rec {
     sed -i -e "s/^CFLAGS = \(.*\)/CFLAGS = \1 -std=gnu89/" Makefile.in
 
     substituteInPlace ./arch/linux/common/modules.h --replace /sbin/modinfo modinfo
-    substituteInPlace ./arch/linux/common/os.h --replace /lib/libc.so.6 ${stdenv.glibc.out}/lib/libc.so.6
+    substituteInPlace ./arch/linux/common/os.h --replace /lib/libc.so.6 ${glibc.out}/lib/libc.so.6
```
Comment (Mindavi) [inline pkgs/tools/system/hardinfo/default.nix:33]: Also.maybe possible to drop the `.6`?

**Review by Mindavi — APPROVED** [review]
Nothing stands out to me, and at least it's not a regression if you fail to replace glibc with stdenv.cc.libc in some places, I guess.

Some small comments, but no blockers.

**Comment by Artturin** [comment]
not sure what to change this to https://github.com/NixOS/nixpkgs/blob/197681059f0fff10d7abd75d11f36654ea063e92/pkgs/stdenv/adapters.nix#L72

File: pkgs/applications/editors/vscode/extensions/cpptools/missing_elf_deps.sh [hunk 4]
```diff
@@ -11,7 +11,7 @@ function get_pkg_out() {
   echo "$(nix-build -E "$nixExp" --no-out-link)${suffix}"
 }
 
-interpreter="$(get_pkg_out "stdenv.glibc" "/lib/ld-linux-x86-64.so.2")"
+interpreter="$(get_pkg_out "glibc" "/lib/ld-linux-x86-64.so.2")"
```
Comment (Ericson2314) [inline pkgs/applications/editors/vscode/extensions/cpptools/missing_elf_deps.sh:14]:
```suggestion
interpreter="$(get_pkg_out "stdenv.cc.libc" "/lib/ld-linux-x86-64.so.2")"
```

File: pkgs/applications/audio/baudline/default.nix [hunk 5]
```diff
@@ -1,5 +1,5 @@
 { lib, stdenv, fetchurl, libXmu, libXt, libX11, libXext, libXxf86vm, libjack2
-, makeWrapper
+, makeWrapper, glibc
```
Comment (Ericson2314) [inline pkgs/applications/audio/baudline/default.nix:2]:
```suggestion
, makeWrapper
```

File: pkgs/applications/audio/baudline/default.nix [hunk 6]
```diff
@@ -36,7 +36,7 @@ stdenv.mkDerivation rec {
 
     cp -r . "$out/libexec/baudline/"
 
-    interpreter="$(echo ${stdenv.glibc.out}/lib/ld-linux*)"
+    interpreter="$(echo ${glibc.out}/lib/ld-linux*)"
```
Comment (Ericson2314) [inline pkgs/applications/audio/baudline/default.nix:39]:
```suggestion
    interpreter="$(echo ${stdenv.cc.libc.out}/lib/ld-linux*)"
```

**Comment by Artturin** [comment]
changed to  lib.warn

**Comment by Artturin** [comment]
please check if the latest commit is correct

**Comment by roberth** [comment]
Removed my review request. I'm not much of an authority on stdenv. I don't see a problem but you really want @Ericson2314 to look at it instead.

**Comment by Artturin** [comment]
fix'd up the commit that did this 

```diff
      extraBuildInputs = (old.extraBuildInputs or []) ++ [
-        stdenv0.glibc.static
+        pkgs.glibc.static
      ];

File: pkgs/development/python-modules/twisted/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/twisted/default.nix:57]: @Artturin I think you missed this

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
Due to the removal of some ``.6`` we probably want to target staging now.

**Comment by Artturin** [comment]
i'd prefer no to so that this can be in 22.05 at release

**Pull request merged**